File: AutoRest/Generators/NodeJS/NodeJS.Tests/AcceptanceTests/validationTests.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

'use strict';

import should = require('should');

import http = require('http');
import util = require('util');
import assert = require('assert');
import msRest = require('ms-rest');
import fs = require('fs');

import stream = require('stream');

import validationClient = require('../Expected/AcceptanceTests/Validation/autoRestValidationTest');

var dummyToken = 'dummy12321343423';
var credentials = new msRest.TokenCredentials(dummyToken);
var clientOptions: msRest.ServiceClientOptions = {};
var baseUri = 'http://localhost:3000';
describe('nodejs', function () {
  describe('Swagger BAT Validation', function () {
    var testClient = new validationClient("abc123", "12-34-5678", baseUri, clientOptions);
    describe('Of Method Parameters', function () {
      it('should test the minimum length constraint on resourceGroupName', function (done) {
        testClient.validationOfMethodParameters("1", 100, function (err, result) {
          should.exist(err);
          err.should.match(/.*resourceGroupName.*constraint.*MinLength.*3/ig);
          done();
        });
      });

      it('should test the maximum length constraint on resourceGroupName', function (done) {
        testClient.validationOfMethodParameters("1234567890A", 100, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*resourceGroupName.*constraint.*MaxLength.*10/ig);
          done();
        });
      });

      it('should test the pattern length constraint on resourceGroupName', function (done) {
        testClient.validationOfMethodParameters("!@#$", 100, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*resourceGroupName.*constraint.*Pattern.*/ig);
          done();
        });
      });

      it('should test the multipleof constraint on id', function (done) {
        testClient.validationOfMethodParameters("123", 105, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*id.*constraint.*MultipleOf.*10/ig);
          done();
        });
      });

      it('should test the InclusiveMinimum constraint on id', function (done) {
        testClient.validationOfMethodParameters("123", 0, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*id.*constraint.*InclusiveMinimum.*100/ig);
          done();
        });
      });

      it('should test the InclusiveMaximum constraint on id', function (done) {
        testClient.validationOfMethodParameters("123", 2000, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*id.*constraint.*InclusiveMaximum.*1000/ig);
          done();
        });
      });

      it('should test the pattern constraint on apiVersion', function (done) {
        var testClient2 = new validationClient("abc123", "12345", baseUri, clientOptions);
        testClient2.validationOfMethodParameters("123", 150, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*apiVersion.*constraint.*Pattern.*/ig);
          done();
        });
      });
    });

    describe('Of Body Parameters', function () {
      it('should test the ExclusiveMinimum constraint on capacity', function (done) {
        testClient.validationOfBody("123", 150, { body: { capacity: 0, child: {}}}, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*capacity.*constraint.*ExclusiveMinimum.*0.*/ig);
          done();
        });
      });

      it('should test the ExclusiveMaximum constraint on capacity', function (done) {
        testClient.validationOfBody("123", 150, { body: { capacity: 2000, child: {} } }, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*capacity.*constraint.*ExclusiveMaximum.*100.*/ig);
          done();
        });
      });

      it('should test the MaxItems constraint on displayNames', function (done) {
        testClient.validationOfBody("123", 150, { body: { displayNames: ["item1", "item2", "item3", "item4", "item5", "item6", "item7"], child: {} } }, function (err, result) {
          should.exist(err);
          err.message.should.match(/.*display_names.*constraint.*MaxItems.*6.*/ig);
          done();
        });
      });
    });

    describe('Constants', function () {
      it('should work in path', function (done) {
        testClient.getWithConstantInPath(function (err, result) {
          should.not.exist(err);
          done();
        });
      });

      it('should work in body', function (done) {
        testClient.postWithConstantInBody({ body: { child: {} } }, function (err, result) {
          should.not.exist(err);
          done();
        });
      });
    });
  });
});

File: AutoRest/Generators/Python/Python.Tests/AcceptanceTests/validation_tests.py
# --------------------------------------------------------------------------
#
# Copyright (c) Microsoft Corporation. All rights reserved.
#
# The MIT License (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the ""Software""), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#
# --------------------------------------------------------------------------

import unittest
import subprocess
import sys
import isodate
import tempfile
import json
from datetime import date, datetime, timedelta
import os
from os.path import dirname, pardir, join, realpath, sep, pardir

cwd = dirname(realpath(__file__))
root = realpath(join(cwd , pardir, pardir, pardir, pardir, pardir))
sys.path.append(join(root, "ClientRuntimes" , "Python", "msrest"))
log_level = int(os.environ.get('PythonLogLevel', 10))

tests = realpath(join(cwd, pardir, "Expected", "AcceptanceTests"))
sys.path.append(join(tests, "Validation"))

from msrest.serialization import Deserializer
from msrest.exceptions import DeserializationError

from autorestvalidationtest import (
    AutoRestValidationTest,
    AutoRestValidationTestConfiguration)
from autorestvalidationtest.models import Product


class ValidationTests(unittest.TestCase):

    def test_constant_values(self):

        config = AutoRestValidationTestConfiguration(
            "abc123",
            "12-34-5678",
            base_url="http://localhost:3000")
        config.log_level = log_level
        client = AutoRestValidationTest(config)

        client.get_with_constant_in_path()
        product = client.post_with_constant_in_body(body=Product())
        self.assertIsNotNone(product)

    def test_validation(self):

        config = AutoRestValidationTestConfiguration(
            "abc123",
            "12-34-5678",
            base_url="http://localhost:3000")
        config.log_level = log_level
        client = AutoRestValidationTest(config)

        with self.assertRaises(ValueError):
            client.validation_of_method_parameters("1", 100)
    #public void ValidationTests()
    #    {
    #        SwaggerSpecRunner.RunTests(
    #            SwaggerPath("validation.json"),
    #            ExpectedPath("Validation"));
    #        var client = new AutoRestValidationTest(Fixture.Uri);
    #        client.SubscriptionId = "abc123";
    #        client.ApiVersion = "12-34-5678";
    #        var exception = Assert.Throws<ValidationException>(() => client.ValidationOfMethodParameters("1", 100));
    #        Assert.Equal(ValidationRules.MinLength, exception.Rule);
    #        Assert.Equal("resourceGroupName", exception.Target);
    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfMethodParameters("1234567890A", 100));
    #        Assert.Equal(ValidationRules.MaxLength, exception.Rule);
    #        Assert.Equal("resourceGroupName", exception.Target);
    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfMethodParameters("!@#$", 100));
    #        Assert.Equal(ValidationRules.Pattern, exception.Rule);
    #        Assert.Equal("resourceGroupName", exception.Target);
    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfMethodParameters("123", 105));
    #        Assert.Equal(ValidationRules.MultipleOf, exception.Rule);
    #        Assert.Equal("id", exception.Target);
    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfMethodParameters("123", 0));
    #        Assert.Equal(ValidationRules.InclusiveMinimum, exception.Rule);
    #        Assert.Equal("id", exception.Target);
    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfMethodParameters("123", 2000));
    #        Assert.Equal(ValidationRules.InclusiveMaximum, exception.Rule);
    #        Assert.Equal("id", exception.Target);

    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfBody("123", 150, new Fixtures.AcceptanceTestsValidation.Models.Product
    #        {
    #            Capacity = 0
    #        }));
    #        Assert.Equal(ValidationRules.ExclusiveMinimum, exception.Rule);
    #        Assert.Equal("Capacity", exception.Target);
    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfBody("123", 150, new Fixtures.AcceptanceTestsValidation.Models.Product
    #        {
    #            Capacity = 100
    #        }));
    #        Assert.Equal(ValidationRules.ExclusiveMaximum, exception.Rule);
    #        Assert.Equal("Capacity", exception.Target);
    #        exception = Assert.Throws<ValidationException>(() => client.ValidationOfBody("123", 150, new Fixtures.AcceptanceTestsValidation.Models.Product
    #        {
    #            DisplayNames = new List<string>
    #            {
    #                "item1","item2","item3","item4","item5","item6","item7"
    #            }
    #        }));
    #        Assert.Equal(ValidationRules.MaxItems, exception.Rule);
    #        Assert.Equal("DisplayNames", exception.Target);

    #        var client2 = new AutoRestValidationTest(Fixture.Uri);
    #        client2.SubscriptionId = "abc123";
    #        client2.ApiVersion = "abc";
    #        exception = Assert.Throws<ValidationException>(() => client2.ValidationOfMethodParameters("123", 150));
    #        Assert.Equal(ValidationRules.Pattern, exception.Rule);
    #        Assert.Equal("ApiVersion", exception.Target);
    #    }